nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKuduProcessor.java

```diff
@@ -0,0 +1,451 @@
+                break;
+            case BINARY:
+                predicate = KuduPredicate.newComparisonPredicate(schema, KuduPredicate.ComparisonOp.valueOf(comparisonOp), Byte.valueOf(value));
+                break;
+            case UNIXTIME_MICROS:
+                predicate = KuduPredicate.newComparisonPredicate(schema, KuduPredicate.ComparisonOp.valueOf(comparisonOp), Integer.valueOf(value));
+                break;
+            case DOUBLE:
+                predicate = KuduPredicate.newComparisonPredicate(schema, KuduPredicate.ComparisonOp.valueOf(comparisonOp), Double.valueOf(value));
+                break;
+            default:
+                throw new IllegalStateException(String.format("unknown type %s", colType));
+        }
+
+        scannerBuilder.addPredicate(predicate);
+    }
+
+    private int getColumnIndex(Schema columns, String colName) {
+        try {
+            return columns.getColumnIndex(colName);
+        } catch (Exception ex) {
+            return -1;
+        }
+    }
+
+    protected Upsert upsertRecordToKudu(KuduTable kuduTable, Record record, List<String> fieldNames, Boolean ignoreNull) {
+        Upsert upsert = kuduTable.newUpsert();
+        buildPartialRow(kuduTable.getSchema(), upsert.getRow(), record, fieldNames, ignoreNull);
+        return upsert;
+    }
+
+    protected Insert insertRecordToKudu(KuduTable kuduTable, Record record, List<String> fieldNames, Boolean ignoreNull) {
+        Insert insert = kuduTable.newInsert();
+        buildPartialRow(kuduTable.getSchema(), insert.getRow(), record, fieldNames, ignoreNull);
+        return insert;
+    }
+
+    protected Delete deleteRecordToKudu(KuduTable kuduTable, Record record, List<String> fieldNames, Boolean ignoreNull) {
+        Delete delete = kuduTable.newDelete();
+        buildPartialRow(kuduTable.getSchema(), delete.getRow(), record, fieldNames, ignoreNull);
+        return delete;
+    }
+
+    protected Update updateRecordToKudu(KuduTable kuduTable, Record record, List<String> fieldNames, Boolean ignoreNull) {
+        Update update = kuduTable.newUpdate();
+        buildPartialRow(kuduTable.getSchema(), update.getRow(), record, fieldNames, ignoreNull);
+        return update;
+    }
+
+    /**
+     * Serializes a row from Kudu to a JSON document of the form:
+     *
+     * {
+     *    "rows": [
+     *      {
+     *          "columnname-1" : "value1",
+     *          "columnname-2" : "value2",
+     *          "columnname-3" : "value3",
+     *          "columnname-4" : "value4",
+     *      },
+     *      {
+     *          "columnname-1" : "value1",
+     *          "columnname-2" : "value2",
+     *          "columnname-3" : "value3",
+     *          "columnname-4" : "value4",
+     *      }
+     *    ]
+     * }
+     */
+    protected String convertToJson(Iterator<RowResult> rows) {
+        final StringBuilder jsonBuilder = new StringBuilder();
```

Comment:
Is this the style other reading type processors use to build JSON? Is there a library that is more elegant than appending strings with StringBuilder?

Reply:
Thinking about this a bit more, but not researching it. Would it be easier and more consistent to convert to a Nifi Record and use Nifi to convert to the needed formats?

Reply:
@granthenke I have just followed the Hbase, Druid Processor's